src/server/game/Anticheat/AnticheatMgr.cpp

```diff
@@ -0,0 +1,453 @@
+        return true;
+    }
+
+    return false;
+}
+
+bool TeleportPlaneCheck::OnCheck(Player* player, MovementInfo const& movementInfo, uint16 /*opcode = 0*/) const
+{
+    if (player->m_movementInfo.pos.GetPositionZ() != 0 ||
+        movementInfo.pos.GetPositionZ() != 0)
+        return false;
+
+    if (movementInfo.HasMovementFlag(MOVEMENTFLAG_FALLING))
+        return false;
+
+    //DEAD_FALLING was deprecated
+    //if (player->getDeathState() == DEAD_FALLING)
+    //    return;
+    float x, y, z;
+    player->GetPosition(x, y, z);
+    float ground_Z = player->GetMap()->GetHeight(x, y, z);
+    float z_diff = fabs(ground_Z - z);
+
+    // we are not really walking there
+    if (z_diff > 1.0f)
+    {
+        TC_LOG_TRACE("entities.player.character", "AnticheatMgr:: Teleport To Plane - Hack detected on player %s (%s)", player->GetName().c_str(), player->GetGUID().ToString().c_str());
+        return true;
+    }
+
+    return false;
+}
+
+// basic detection
+bool ClimbHackCheck::OnCheck(Player* player, MovementInfo const& movementInfo, uint16 opcode /*= 0*/) const
+{
+    AnticheatData* playerData = GetPlayerData(player->GetGUID());
+    if (!playerData)
+        return false;
+
+    if (opcode != MSG_MOVE_HEARTBEAT || playerData->GetLastOpcode() != MSG_MOVE_HEARTBEAT)
+        return false;
+
+    // in this case we don't care if they are "legal" flags, they are handled in another parts of the Anticheat Manager.
+    if (player->IsInWater() ||
+        player->IsFlying() ||
+        player->IsFalling())
+        return false;
+
+    Position playerPos = player->GetPosition();
+
+    float deltaZ = fabsf(playerPos.GetPositionZ() - movementInfo.pos.GetPositionZ());
+    float deltaXY = movementInfo.pos.GetExactDist2d(&playerPos);
+
+    // Prevent divide by 0
+    if (G3D::fuzzyEq(deltaXY, 0.f))
+        return false;
+
+    float angle = Position::NormalizeOrientation(tangent(deltaZ / deltaXY));
+
+    if (angle > CLIMB_ANGLE)
+    {
+        TC_LOG_TRACE("entities.player.character", "AnticheatMgr:: Climb-Hack detected on player %s (%s)", player->GetName().c_str(), player->GetGUID().ToString().c_str());
+        return true;
+    }
+
+    return false;
+}
+
+void AnticheatMgr::StartHackDetection(Player* player, MovementInfo const& movementInfo, uint16 opcode)
+{
+    if (!m_enabled)
+        return;
+
+    // Disable anticheat for Game Masters
+    if (player->CanBeGameMaster())
+        return;
+
+    AnticheatData& plData = m_playerMap[player->GetGUID()];
+    if (player->IsInFlight() || player->GetTransport() || player->GetVehicle())
+    {
+        plData.SetLastOpcode(opcode);
+        return;
+    }
+
+    for (AnticheatCheck* check : m_checks)
+        if (check->OnCheck(player, movementInfo, opcode))
+            check->HackReport(player);
+
+    plData.SetLastOpcode(opcode);
+}
+
+void AnticheatMgr::HandlePlayerLogin(Player* player, PreparedQueryResult result)
+{
+    AnticheatData& plData = m_playerMap[player->GetGUID()];
```

Comment:
the whole idea of this line in case result is null is to create the AnticheatData element in the container ?